### ISSUE TITLE
pymc-client.py fails on window to sample

### ISSUE THREAD

**Issue by krosenfeld-IDM** [UM-Bridge/umbridge#7]
I am running the `pymc-client.py` [script](https://github.com/UM-Bridge/umbridge/blob/main/clients/python/umbridge-client.py) using the analytic gaussian mixture benchmark:

```
docker run -it -p 4243:4243 linusseelinger/benchmark-analytic-gaussian-mixture
python3 umbridge/clients/python pymc-client.py http://localhost:4243
```

and seeing an error during `pm.sample`:

```
http://localhost:4243
Connecting to host URL http://localhost:4243
Model output: [-48.17723351]
Check model's gradient against numerical gradient. This requires an UM-Bridge model with gradient support.
 |----------------------------------------------------------| 0.02% [1/5000 00:00<? logp = -3.7296, ||grad|| = 0.025167] |----------------------------------------------------------| 0.04% [2/5000 00:00<? logp = -3.7296, ||grad|| = 0.025167] |----------------------------------------------------------| 0.06% [3/5000 00:00<? logp = -3.7296, ||grad|| = 0.025167] |----------------------------------------------------------| 0.08% [4/5000 00:00<? logp = -3.7296, ||grad|| = 0.025167] |----------------------------------------------------------| 0.10% [5/5000 00:00<? logp = -3.7296, ||grad|| = 0.025167] |----------------------------------------------------------| 0.12% [6/5000 00:00<? logp = -3.7296, ||grad|| = 0.025167] |----------------------------------------------------------| 0.14% [7/5000 00:00<? logp = -3.7296, ||grad|| = 0.025167] |----------------------------------------------------------| 0.16% [8/5000 00:00<? logp = -3.7296, ||grad|| = 0.025167] |███████████████████████████████████████████████████████| 100.00% [8/8 00:00<00:00 logp = -3.7296, ||grad|| = 0.025167]
MAP estimate of posterior is [-1.99936594  1.99936594]
Only 50 samples in chain.
Auto-assigning NUTS sampler...
Initializing NUTS using jitter+adapt_diag...
Multiprocess sampling (4 chains in 4 jobs)
NUTS: [posterior]
Connecting to host URL http://localhost:4243
Model output: [-48.17723351]
Check model's gradient against numerical gradient. This requires an UM-Bridge model with gradient support.
 |----------------------------------------------------------| 0.02% [1/5000 00:00<? logp = -3.7296, ||grad|| = 0.025167] |----------------------------------------------------------| 0.04% [2/5000 00:00<? logp = -3.7296, ||grad|| = 0.025167] |----------------------------------------------------------| 0.06% [3/5000 00:00<? logp = -3.7296, ||grad|| = 0.025167] |----------------------------------------------------------| 0.08% [4/5000 00:00<? logp = -3.7296, ||grad|| = 0.025167] |----------------------------------------------------------| 0.10% [5/5000 00:00<? logp = -3.7296, ||grad|| = 0.025167] |----------------------------------------------------------| 0.12% [6/5000 00:00<? logp = -3.7296, ||grad|| = 0.025167] |----------------------------------------------------------| 0.14% [7/5000 00:00<? logp = -3.7296, ||grad|| = 0.025167] |----------------------------------------------------------| 0.16% [8/5000 00:00<? logp = -3.7296, ||grad|| = 0.025167] |███████████████████████████████████████████████████████| 100.00% [8/8 00:00<00:00 logp = -3.7296, ||grad|| = 0.025167]
MAP estimate of posterior is [-1.99936594  1.99936594]
Only 50 samples in chain.
Auto-assigning NUTS sampler...
Initializing NUTS using jitter+adapt_diag...
Multiprocess sampling (4 chains in 4 jobs)
NUTS: [posterior]
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "C:\Users\katherinero\Documents\code\umbridge\env\lib\multiprocessing\spawn.py", line 116, in spawn_main
    exitcode = _main(fd, parent_sentinel)
  File "C:\Users\katherinero\Documents\code\umbridge\env\lib\multiprocessing\spawn.py", line 125, in _main
    prepare(preparation_data)
  File "C:\Users\katherinero\Documents\code\umbridge\env\lib\multiprocessing\spawn.py", line 236, in prepare
    _fixup_main_from_path(data['init_main_from_path'])
  File "C:\Users\katherinero\Documents\code\umbridge\env\lib\multiprocessing\spawn.py", line 287, in _fixup_main_from_path
    main_content = runpy.run_path(main_path,
  File "C:\Users\katherinero\Documents\code\umbridge\env\lib\runpy.py", line 265, in run_path
    return _run_module_code(code, init_globals, run_name,
  File "C:\Users\katherinero\Documents\code\umbridge\env\lib\runpy.py", line 97, in _run_module_code
    _run_code(code, mod_globals, init_globals,
  File "C:\Users\katherinero\Documents\code\umbridge\env\lib\runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "C:\Users\katherinero\Documents\code\umbridge\clients\python\pymc-client.py", line 40, in <module>
    inferencedata = pm.sample(draws=50)
  File "C:\Users\katherinero\Documents\code\umbridge\env\lib\site-packages\pymc\sampling.py", line 617, in sample
    mtrace = _mp_sample(**sample_args, **parallel_args)
  File "C:\Users\katherinero\Documents\code\umbridge\env\lib\site-packages\pymc\sampling.py", line 1508, in _mp_sample
    sampler = ps.ParallelSampler(
  File "C:\Users\katherinero\Documents\code\umbridge\env\lib\site-packages\pymc\parallel_sampling.py", line 412, in __init__
    self._samplers = [
  File "C:\Users\katherinero\Documents\code\umbridge\env\lib\site-packages\pymc\parallel_sampling.py", line 413, in <listcomp>
    ProcessAdapter(
  File "C:\Users\katherinero\Documents\code\umbridge\env\lib\site-packages\pymc\parallel_sampling.py", line 272, in __init__
    self._process.start()
  File "C:\Users\katherinero\Documents\code\umbridge\env\lib\multiprocessing\process.py", line 121, in start
    self._popen = self._Popen(self)
  File "C:\Users\katherinero\Documents\code\umbridge\env\lib\multiprocessing\context.py", line 327, in _Popen
    return Popen(process_obj)
  File "C:\Users\katherinero\Documents\code\umbridge\env\lib\multiprocessing\popen_spawn_win32.py", line 45, in __init__
    prep_data = spawn.get_preparation_data(process_obj._name)
  File "C:\Users\katherinero\Documents\code\umbridge\env\lib\multiprocessing\spawn.py", line 154, in get_preparation_data
    _check_not_importing_main()
  File "C:\Users\katherinero\Documents\code\umbridge\env\lib\multiprocessing\spawn.py", line 134, in _check_not_importing_main
    raise RuntimeError('''
RuntimeError:
        An attempt has been made to start a new process before the
        current process has finished its bootstrapping phase.

        This probably means that you are not using fork to start your
        child processes and you have forgotten to use the proper idiom
        in the main module:

            if __name__ == '__main__':
                freeze_support()
                ...

        The "freeze_support()" line can be omitted if the program
        is not going to be frozen to produce an executable.
```

**Comment by krosenfeld-IDM** [comment]
System is python 3.8  (anaconda) in windows 10

**Comment by krosenfeld-IDM** [comment]
if I add `if __name__ == "__main__"` then the code runs further and then errors out:

```
'(env)'C:\Users\katherinero\Documents\code\umbridge\clients\python>python pymc-client.py http://localhost:4243
Connecting to host URL http://localhost:4243
Model output: [-48.17723351]
Check model's gradient against numerical gradient. This requires an UM-Bridge model with gradient support.
 |----------------------------------------------------------| 0.02% [1/5000 00:00<? logp = -3.7296, ||grad|| = 0.025167] |----------------------------------------------------------| 0.04% [2/5000 00:00<? logp = -3.7296, ||grad|| = 0.025167] |----------------------------------------------------------| 0.06% [3/5000 00:00<? logp = -3.7296, ||grad|| = 0.025167] |----------------------------------------------------------| 0.08% [4/5000 00:00<? logp = -3.7296, ||grad|| = 0.025167] |----------------------------------------------------------| 0.10% [5/5000 00:00<? logp = -3.7296, ||grad|| = 0.025167] |----------------------------------------------------------| 0.12% [6/5000 00:00<? logp = -3.7296, ||grad|| = 0.025167] |----------------------------------------------------------| 0.14% [7/5000 00:00<? logp = -3.7296, ||grad|| = 0.025167] |----------------------------------------------------------| 0.16% [8/5000 00:00<? logp = -3.7296, ||grad|| = 0.025167] |███████████████████████████████████████████████████████| 100.00% [8/8 00:00<00:00 logp = -3.7296, ||grad|| = 0.025167]
MAP estimate of posterior is [-1.99936594  1.99936594]
Only 50 samples in chain.
Auto-assigning NUTS sampler...
Initializing NUTS using jitter+adapt_diag...
Multiprocess sampling (4 chains in 4 jobs)
NUTS: [posterior]
 |------------------------------------------------------------------------| 0.02% [1/4200 00:00<23:28 Sampling 4 chains, 0 divergences] |------------------------------------------------------------------------| 0.05% [2/4200 00:00<15:17 Sampling 4 chains, 0 divergences] |------------------------------------------------------------------------| 0.07% [3/4200 00:00<14:32 Sampling 4 chains, 0 divergences] |------------------------------------------------------------------------| 0.10% [4/4200 00:00<11:50 Sampling 4 chains, 0 divergences] |------------------------------------------------------------------------| 0.12% [5/4200 00:00<09:42 Sampling 4 chains, 0 divergences] |------------------------------------------------------------------------| 0.14% [6/4200 00:00<08:12 Sampling 4 chains, 0 divergences] |------------------------------------------------------------------------| 0.17% [7/4200 00:00<07:33 Sampling 4 chains, 0 divergences] |------------------------------------------------------------------------| 0.19% [8/4200 00:00<06:51 Sampling 4 chains, 0 divergences]C:\Users\katherinero\Documents\code\umbridge\env\lib\site-packages\aesara\link\utils.py:529: UserWarning: <class 'requests.exceptions.JSONDecodeError'> error does not allow us to add an extra error message
  warnings.warn(
Traceback (most recent call last):
  File "pymc-client.py", line 42, in <module>
    inferencedata = pm.sample(draws=50)
  File "C:\Users\katherinero\Documents\code\umbridge\env\lib\site-packages\pymc\sampling.py", line 617, in sample
    mtrace = _mp_sample(**sample_args, **parallel_args)
  File "C:\Users\katherinero\Documents\code\umbridge\env\lib\site-packages\pymc\sampling.py", line 1522, in _mp_sample
    for draw in sampler:
  File "C:\Users\katherinero\Documents\code\umbridge\env\lib\site-packages\pymc\parallel_sampling.py", line 458, in __iter__
    draw = ProcessAdapter.recv_draw(self._active)
  File "C:\Users\katherinero\Documents\code\umbridge\env\lib\site-packages\pymc\parallel_sampling.py", line 333, in recv_draw
    msg = ready[0].recv()
  File "C:\Users\katherinero\Documents\code\umbridge\env\lib\multiprocessing\connection.py", line 251, in recv
    return _ForkingPickler.loads(buf.getbuffer())
  File "C:\Users\katherinero\Documents\code\umbridge\env\lib\site-packages\requests\exceptions.py", line 41, in __init__
    CompatJSONDecodeError.__init__(self, *args)
TypeError: __init__() missing 2 required positional arguments: 'doc' and 'pos'
```

This error goes away if I use the Metropolis sampler.

```
        inferencedata = pm.sample(draws=50, step=pm.Metropolis())
```

**Comment by linusseelinger** [comment]
I think that's a PyMC issue when running in parallel on windows (see [pymc discourse](https://discourse.pymc.io/t/error-during-run-sampling-method/2522/5)). UM-Bridge does use multiprocessing, but that is only the case for the server/model side. Do the proposed fixes from that thread (setting `cores=1` or wrapping only the `sample(...)` call in `if __name__ == "__main__"`) help in your case?

I could imagine that PyMC's Metropolis sampler is sequential by default, maybe that's why it runs.

Separate from that issue: More than cores=1 probably won't help if you are connecting to a single instance of an umbridge model. With parallel instances of umbridge models + load balancer (like we are doing in our kubernetes reference setup for cloud applications), the same code with cores>1 should immediately provide gains in performance.

**Comment by krosenfeld-IDM** [comment]
The error is not resolved by wrapping `sample(...)` in `if __name__ === "__main__":`  or `cores=1`.  However it does sometimes finish - something going on with the stochastic nature of the sampler.  

It might be an issue with the benchmark model itself. If I try to evaluate the posterior at a value far from the location of the gaussians (no pymc at all): 
```
import umbridge
model = umbridge.HTTPModel("http://localhost:4243", "posterior")
print(model([[100.0, 18.0]]))
```

I see the `JSONDecodeError` error:
```
'(env)'C:\Users\katherinero\Documents\code\umbridge\clients\python>python basic_eval.py
Traceback (most recent call last):
  File "C:\Users\katherinero\Documents\code\umbridge\env\lib\site-packages\requests\models.py", line 971, in json
    return complexjson.loads(self.text, **kwargs)
  File "C:\Users\katherinero\Documents\code\umbridge\env\lib\json\__init__.py", line 357, in loads
    return _default_decoder.decode(s)
  File "C:\Users\katherinero\Documents\code\umbridge\env\lib\json\decoder.py", line 337, in decode
    obj, end = self.raw_decode(s, idx=_w(s, 0).end())
  File "C:\Users\katherinero\Documents\code\umbridge\env\lib\json\decoder.py", line 355, in raw_decode
    raise JSONDecodeError("Expecting value", s, err.value) from None
json.decoder.JSONDecodeError: Expecting value: line 1 column 14 (char 13)

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "basic_eval.py", line 3, in <module>
    print(model([[100.0, 18.0]]))
  File "c:\users\katherinero\documents\code\umbridge\umbridge\um.py", line 96, in __call__
    response = requests.post(f"{self.url}/Evaluate", json=inputParams).json()
  File "C:\Users\katherinero\Documents\code\umbridge\env\lib\site-packages\requests\models.py", line 975, in json
    raise RequestsJSONDecodeError(e.msg, e.doc, e.pos)
requests.exceptions.JSONDecodeError: Expecting value: line 1 column 14 (char 13)
```

But if I evaluate closer to their locations:

```
import umbridge
model = umbridge.HTTPModel("http://localhost:4243", "posterior")
print(model([[1, 1]]))
```

successfully  outputs `[[-1.9265813564722167]]`.

Probably an issue with evaluating the log of a really small number and generating a `RuntimeWarning`.

And thanks @linusseelinger for the note regarding using parallel instances of umbridge models with load balancing.  I think I'm *almost* ready with a first server example for our model and might need to look into speeding up client evaluation (or just hand off to @annereinarz  but I'd like to know it is being somewhat sensical first...).

**Comment by linusseelinger** [comment]
Should be fixed by #8 . Don't think we can do something about the gaussian mixture.

Thanks a lot for reporting!
